### PR TITLE
(RE-5817) Add foss-only mode

### PR DESCRIPTION
**Pull request by underscorgan**
This updates the retrieve jenkins task to only fetch specified platforms
if FOSS_ONLY is set to true. It will default to using foss_platforms if
specified, but that can be overwritten with the FOSS_PLATFORMS
environment variable which should be a space-separated list of
platform-tags.

### CODE REVIEW

**Comment by underscorgan** [comment]
Still to-do: 

- [ ] Add some deduplication (will currently fetch all packages for the versions once per architecture, we could be smarter about this)
- [ ] Add more robustness / print warnings when something goes wrong and we want to fail back
- [ ] Add some spec testing maybe????

File: lib/packaging/config/params.rb [hunk 1]
```diff
@@ -172,6 +173,8 @@ module Pkg::Params
               { :var => :default_cow,             :envvar => :COW },
               { :var => :default_mock,            :envvar => :MOCK },
               { :var => :final_mocks,             :envvar => :MOCK },
+              { :var => :foss_only,               :envvar => :FOSS_ONLY },
+              { :var => :foss_platforms,          :envvar => :FOSS_PLATFORMS,  :type => :array },
```
Comment (mckern) [inline lib/packaging/config/params.rb]: Yaaaassssssssss `:type => :array`! I knew that was going to be handy.

File: spec/lib/packaging/util/platform_spec.rb [hunk 2]
```diff
@@ -32,6 +32,19 @@
     end
   end
 
+  describe '#artifacts_path' do
+    it 'should be correct' do
+      expect(Pkg::Util::Platform.artifacts_path('el-7-x86_64')).to eq('artifacts/el/7')
```
Comment (mckern) [inline spec/lib/packaging/util/platform_spec.rb]: Would we benefit from testing other platforms?

Comment (underscorgan) [inline spec/lib/packaging/util/platform_spec.rb]: Probably. I was slightly lazy / copypastad the tests, but I can udpate it to hit all the cases in the case statement.

Comment (underscorgan) [inline spec/lib/packaging/util/platform_spec.rb]: More tests added!

File: tasks/retrieve.rake [hunk 3]
```diff
@@ -17,17 +17,45 @@ namespace :pl do
...
-        # For the next person who needs to look these flags up:
-        # -r = recursive
-        # -l 0 = infinitely recurse, no limit
-        # --cut-dirs 3 = will cut off #{Pkg::Config.project}, #{Pkg::Config.ref}, and the first directory in #{remote_target} from the url when saving to disk
-        # -np = Only descend when recursing, never ascend
-        # -nH = Discard http://#{Pkg::Config.builds_server} when saving to disk
-        # --reject = Reject all hits that match the supplied regex
-        # -P = where to save to disk (defaults to ./)
-        sh "#{wget} -r -np -nH -l 0 --cut-dirs 3 -P #{local_target} --reject 'index*' #{package_url}/"
+        if Pkg::Config.foss_only and not Pkg::Config.foss_platforms
```
Comment (mckern) [inline tasks/retrieve.rake]: `&& !` is [probably better than](http://phrogz.net/ProgrammingRuby/language.html#table_18.4) `and not`. Wheeeeee!

File: lib/packaging/util/platform.rb [hunk 4]
```diff
@@ -41,6 +41,28 @@ def get_attribute(platform_tag, attribute_name)
       end
     end
 
+    def artifacts_path(platform_tag)
```
Comment (melissa) [inline lib/packaging/util/platform.rb]: This makes me sad... I wish we didn't have to define this everywhere

File: lib/packaging/util/platform.rb [hunk 5]
```diff
@@ -41,6 +41,28 @@ def get_attribute(platform_tag, attribute_name)
...
+        File.join('artifacts', 'deb', get_attribute(platform_tag, :codename))
+      when 'svr4', 'ips'
+        # solaris/10/PC1 for example
+        File.join('artifacts', 'solaris', version)
+      when 'dmg'
+        File.join('artifacts', 'apple', version)
+      when 'msi'
+        File.join('artifacts', 'windows')
+      else
+        fail "Not sure what to do with a package format of '#{package_format}'"
```
Comment (melissa) [inline lib/packaging/util/platform.rb]: Maybe this should be `"Not sure where to find packages that belong to the format type '#{package_format}'"` to be more explicit with what's failing?

File: lib/packaging/util/platform.rb [hunk 6]
```diff
@@ -41,6 +41,28 @@ def get_attribute(platform_tag, attribute_name)
...
+      when 'rpm', 'swix'
+        # el/7/PC1/x86_64 for example
+        File.join('artifacts', platform, version)
+      when 'deb'
+        File.join('artifacts', 'deb', get_attribute(platform_tag, :codename))
+      when 'svr4', 'ips'
+        # solaris/10/PC1 for example
+        File.join('artifacts', 'solaris', version)
+      when 'dmg'
+        File.join('artifacts', 'apple', version)
```
Comment (melissa) [inline lib/packaging/util/platform.rb]: soooo, this is a vanagon thing. It looks like dmgs that are built with the packaging repo don't have the version in the path (`artifacts/apple/puppet-3.8.6.dmg` )

File: lib/packaging/util/platform.rb (hunk 4 above)
Comment (underscorgan) [inline lib/packaging/util/platform.rb]: I know :\

It's slightly different than the repo_path though so this seemed cleanest?

File: lib/packaging/util/platform.rb (hunk 6 above)
Comment (melissa) [inline lib/packaging/util/platform.rb]: Luckily we don't have the same issue with windows

File: lib/packaging/util/platform.rb (hunk 4 above)
Comment (melissa) [inline lib/packaging/util/platform.rb]: It probably is :disappointed:

File: tasks/retrieve.rake [hunk 7]
```diff
@@ -17,17 +17,45 @@ namespace :pl do
...
-        # -P = where to save to disk (defaults to ./)
-        sh "#{wget} -r -np -nH -l 0 --cut-dirs 3 -P #{local_target} --reject 'index*' #{package_url}/"
+        if Pkg::Config.foss_only && !Pkg::Config.foss_platforms
+          warn "FOSS_ONLY specified, but I don't know anything about FOSS_PLATFORMS. Fetch everything?"
+          unless ask_yes_or_no
+            warn "Retrieve cancelled"
+            exit
+          end
+        elsif Pkg::Config.foss_only && remote_target != 'artifacts'
+          warn "I only know how to fetch from remote_target 'artifacts' with FOSS_ONLY. Fetch everything?"
```
Comment (melissa) [inline tasks/retrieve.rake]: Let's expand the `Fetch everything?` to be `Fetch everything from artifacts?` to clarify that they're not going to get anything from whatever the remote target is

Comment (melissa) [inline tasks/retrieve.rake]: wait, nvm, I totally misinterpreted what's happening with this. ignore me

Comment (underscorgan) [inline tasks/retrieve.rake]: If they answer yes it'll fall back to default behavior so it won't fetch from artifacts, but from whatever target they specified.

**Comment by melissa** [comment]
Other than the issue with the dmgs and this needing a rebase, I'm :+1:

**Comment by underscorgan** [comment]
I've got a dmg yak to shave here. Will look again in the morning.

**Comment by underscorgan** [comment]
This should be good to go now!

**Comment by melissa** [comment]
@underscorgan does it blend?

**Comment by underscorgan** [comment]
@melissa YEP. I've tested it locally (yay for easy to test things!), and I can fetch everything, fetch overridden platforms, fetch for puppet and puppet-agent (to test the osx oddities).

File: lib/packaging/util/net.rb [hunk 8]
```diff
@@ -124,6 +124,17 @@ def curl_form_data(uri, form_data = [], options = {})
...
 
+    def uri_status_code(uri)
+      data = [
+        '--request GET',
+        '--silent',
+        '--location',
+        '--write-out "%{http_code}"',
+        '--output /dev/null'
+      ]
+      Pkg::Util::Net.curl_form_data(uri, data)
```
Comment (mckern) [inline lib/packaging/util/net.rb]: kind of love the abuse of `#curl_form_data`.

Comment (underscorgan) [inline lib/packaging/util/net.rb]: it was already there, so may as well reuse it :)

**Comment by melissa** [comment]
:100:

**Pull request merged**